agent/internal/detect.go

```diff
@@ -8,35 +8,49 @@ import (
 	"strconv"
 	"strings"
 
+	"github.com/google/uuid"
+
 	"github.com/pkg/errors"
 	"github.com/shirou/gopsutil/cpu"
 	log "github.com/sirupsen/logrus"
 
 	"github.com/determined-ai/determined/master/pkg/device"
 )
 
-// detectDevices returns a slice of Device's representing the devices exposed by the agent.
-// Autoconfigure the devices exposed by the agent.
-//
-// The most common case in deployed installs is to expose all GPU devices present. To support
-// various testing configurations, we also allow the agent to expose fake devices, a subset of CPU
-// resources or a subset of GPU resources, but this is not representative of deployed agents.
-//
-// The current policy is:
-// - Expose all GPUs present on the machine.
-// - If there are no GPUs, expose all CPUs present on the machine after applying the optional mask
-// `cpu_limit`.
-//
-// An error is returned instead if detection failed unexpectedly.
-func detectDevices(visibleGPUs string) ([]device.Device, error) {
-	switch devices, err := detectGPUs(visibleGPUs); {
-	case err != nil:
-		return nil, errors.Wrap(err, "error while gathering GPU info through nvidia-smi command")
-	case len(devices) != 0:
-		return devices, nil
+func (a *agent) detect() error {
+	switch {
+	case a.ArtificialSlots > 0:
+		for i := 0; i < a.ArtificialSlots; i++ {
+			id := uuid.New().String()
+			a.Devices = append(a.Devices, device.Device{
+				ID: i, Brand: "Artificial", UUID: id, Type: device.CPU})
+		}
+	case a.SlotType == "gpu":
+		devices, err := detectGPUs(a.Options.VisibleGPUs)
+		if err != nil {
+			return errors.Wrap(err, "error while gathering GPU info through nvidia-smi command")
+		}
+		a.Devices = devices
```

Comment:
blocking: If `gpu` is set and `len(a.Devices) == 0` should we return an error instead of 0 devices?

Reply:
we should return 0 devices support the case there are no GPUs for zero-slot resource pools. It should only return an error if it finds `nvidia-smi` returns an error.